stetho/src/main/java/com/facebook/stetho/screencast/android/ScreencastDispatcher.java

```diff
@@ -0,0 +1,150 @@
+  private Page.StartScreencastRequest mRequest;
+  private final RectF mTempSrc = new RectF();
+  private final RectF mTempDst = new RectF();
+
+  public ScreencastDispatcher() {
+  }
+
+  public void startScreencast(JsonRpcPeer peer, Page.StartScreencastRequest request) {
+    mRequest = request;
+
+    LogUtil.d("Starting screencast");
+    if (mHandlerThread == null) {
+      mHandlerThread = new HandlerThread("Screencast Thread");
+      mHandlerThread.start();
+    }
+    mPeer = peer;
+    mIsRunning = true;
+    mBackgroundHandler = new Handler(mHandlerThread.getLooper());
+    mMainHandler.postDelayed(mBitmapFetchRunnable, FRAME_DELAY);
+  }
+
+  public void stopScreencast() {
+    LogUtil.d("Stopping screencast");
+    mMainHandler.removeCallbacks(mBitmapFetchRunnable);
+    mBackgroundHandler.removeCallbacks(mEventDispatchRunnable);
+    mIsRunning = false;
+    mBitmap = null;
+    mCanvas = null;
+  }
+
+  private class BitmapFetchRunnable implements Runnable {
+    @Override
+    public void run() {
+      getScreenBitmap();
+      mBackgroundHandler.post(mEventDispatchRunnable.withEndAction(this));
+    }
+
+    private Bitmap getScreenBitmap() {
+      if (!mIsRunning) {
+        return null;
+      }
+      Activity activity = mActivityTracker.tryGetTopActivity();
+      if (activity == null) {
+        return null;
+      }
+      // This stuff needs to happen in the UI thread
+      View rootView = activity.getWindow().getDecorView();
+      try {
+        if (mBitmap == null) {
+          int viewWidth = rootView.getWidth();
+          int viewHeight = rootView.getHeight();
+          float scale = Math.min((float) mRequest.maxWidth / (float) viewWidth,
+              (float) mRequest.maxHeight / (float) viewHeight);
+          int destWidth = (int) (viewWidth * scale);
+          int destHeight = (int) (viewHeight * scale);
+          mBitmap = Bitmap.createBitmap(destWidth, destHeight, Bitmap.Config.RGB_565);
+          mCanvas = new Canvas(mBitmap);
+          Matrix matrix = new Matrix();
+          mTempSrc.set(0, 0, viewWidth, viewHeight);
+          mTempDst.set(0, 0, destWidth, destHeight);
+          matrix.setRectToRect(mTempSrc, mTempDst, Matrix.ScaleToFit.CENTER);
+          mCanvas.setMatrix(matrix);
+        }
+        rootView.draw(mCanvas);
+        return mBitmap;
+      } catch (OutOfMemoryError e) {
+        LogUtil.w("Out of memory trying to allocate screencast mBitmap.");
+        return null;
+      }
+    }
+  }
+
+  private class EventDispatchRunnable implements Runnable {
+    private Runnable mEndAction;
+
+    private EventDispatchRunnable withEndAction(Runnable endAction) {
+      mEndAction = endAction;
+      return this;
+    }
+
+    @Override
+    public void run() {
+      if (!mIsRunning || mBitmap == null) {
+        return;
+      }
+      int width = mBitmap.getWidth();
+      int height = mBitmap.getHeight();
+      ByteArrayOutputStream stream = new ByteArrayOutputStream();
```

Comment:
This code creates a ByteArrayOutputStream which houses a growable byte[] buffer. This is then copied to a new array, and then copied into a String. Lots of copying :)

I believe we can reduce allocation and churn by pooling the ByteArrayOutputStream as well and avoiding its toByteArray() method. In order to do that we'll have to make a derived class in order to access the protected 'buf' field (like with a method called getBuffer()). 

If we're willing to go all the way, we can make our serializer understand something other than 'String' for when we do the "event.data" assignment. Perhaps we can make our serializer understand CharSequence? And then we can plug in our own CharSequence implementation that references the byte[] buffer and the length we care about. (I haven't looked at the serializer yet to see how or if this could work without copying ... it would need to avoid accessing the char[] array)